Description: CVE-2025-13699 - Fix directory traversal in mariadb-dump -T
 MariaDB's mariadb-dump utility contains a directory traversal vulnerability
 in view/table name handling (MDEV-37483). When using mariadb-dump -T, crafted
 table/view names containing path separators can enable arbitrary file writes.
 This patch introduces a build_path_for_table() helper function that converts
 table names to filesystem-safe names using the filename charset, and fixes
 mysqlimport to reverse-convert @-encoded filenames back to proper table names.
 Backported from MariaDB 10.6.24 to 10.5.29 for Debian bullseye.
Origin: upstream, https://github.com/MariaDB/server/commit/ff12ec86a5898a5a4a4eeb77be26ecbd711b3128
Bug: https://jira.mariadb.org/browse/MDEV-37483
Author: Sergei Golubchik <serg@mariadb.org>
Last-Update: 2026-02-16

--- a/client/mysqldump.c
+++ b/client/mysqldump.c
@@ -1811,6 +1811,27 @@
   return query_str;
 }
 
+
+static const char* build_path_for_table(char *to, const char *dir,
+                                        const char *table, const char *ext)
+{
+  char filename[FN_REFLEN], tmp_path[FN_REFLEN];
+  convert_dirname(tmp_path, path, NULL);
+  my_load_path(tmp_path, tmp_path, NULL);
+  if (check_if_legal_tablename(table))
+    strxnmov(filename, sizeof(filename) - 1, table, "@@@", NULL);
+  else
+  {
+    uint errors, len;
+    len= my_convert(filename, sizeof(filename) - 1, &my_charset_filename,
+                    table, (uint32)strlen(table), charset_info, &errors);
+    filename[len]= 0;
+  }
+  return fn_format(to, filename, tmp_path, ext, MYF(MY_UNPACK_FILENAME));
+}
+
+
+
 /*
   Open a new .sql file to dump the table or view into
 
@@ -1825,12 +1846,9 @@
 */
 static FILE* open_sql_file_for_table(const char* table, int flags)
 {
-  FILE* res;
-  char filename[FN_REFLEN], tmp_path[FN_REFLEN];
-  convert_dirname(tmp_path,path,NullS);
-  res= my_fopen(fn_format(filename, table, tmp_path, ".sql", 4),
-                flags, MYF(MY_WME));
-  return res;
+  char filename[FN_REFLEN];
+  return my_fopen(build_path_for_table(filename, path, table, ".sql"),
+                  flags, MYF(MY_WME));
 }
 
 
@@ -4017,15 +4035,9 @@
 
   if (path)
   {
-    char filename[FN_REFLEN], tmp_path[FN_REFLEN];
+    char filename[FN_REFLEN];
 
-    /*
-      Convert the path to native os format
-      and resolve to the full filepath.
-    */
-    convert_dirname(tmp_path,path,NullS);    
-    my_load_path(tmp_path, tmp_path, NULL);
-    fn_format(filename, table, tmp_path, ".txt", MYF(MY_UNPACK_FILENAME));
+    build_path_for_table(filename, path, table, ".txt");
 
     /* Must delete the file that 'INTO OUTFILE' will write to */
     my_delete(filename, MYF(0));
@@ -4034,7 +4046,6 @@
     to_unix_path(filename);
 
     /* now build the query string */
-
     dynstr_append_checked(&query_string, "SELECT /*!40001 SQL_NO_CACHE */ ");
     dynstr_append_checked(&query_string, select_field_names.str);
     dynstr_append_checked(&query_string, " INTO OUTFILE '");
--- a/client/mysqlimport.c
+++ b/client/mysqlimport.c
@@ -317,7 +317,17 @@
   DBUG_ENTER("write_to_table");
   DBUG_PRINT("enter",("filename: %s",filename));
 
-  fn_format(tablename, filename, "", "", 1 | 2); /* removes path & ext. */
+  fn_format(tablename, filename, "", "", MYF(MY_REPLACE_DIR | MY_REPLACE_EXT));
+  if (strchr(tablename, '@'))
+  {
+    uint errors, len;
+    const char *csname= my_default_csname(); /* see MYSQL_SET_CHARSET_NAME */
+    CHARSET_INFO *cs= get_charset_by_csname(csname,  MY_CS_PRIMARY, MYF(0));
+    len= my_convert(escaped_name, sizeof(escaped_name) - 1, cs, tablename,
+                    (uint32)strlen(tablename), &my_charset_filename, &errors);
+    if (!errors)
+      strmake(tablename, escaped_name, len);
+  }
   if (!opt_local_file)
     strmov(hard_path,filename);
   else
--- a/mysql-test/main/mysqldump.test
+++ b/mysql-test/main/mysqldump.test
@@ -3033,3 +3033,45 @@
 drop table t1;
 
 --echo # End of 10.5 tests
+
+--echo #
+--echo # MDEV-37483 mariadb-dump -T doesn't convert table names
+--echo #
+set names latin1;
+create database foo;
+use foo;
+
+create table `con_schöne_grüße` (a int) select 1 as a;
+create table `con` (b int) select 2 as b;
+create table `con/bar` (c int) select 3 as c;
+create table `con@home` (d int) select 4 as d;
+exec $MYSQL_DUMP foo --tab $MYSQLTEST_VARDIR/tmp;
+drop database foo;
+use test;
+move_file $MYSQLTEST_VARDIR/tmp/con@0040home.sql $MYSQLTEST_VARDIR/tmp/con@home.sql;
+move_file $MYSQLTEST_VARDIR/tmp/con@0040home.txt $MYSQLTEST_VARDIR/tmp/con@home.txt;
+list_files $MYSQLTEST_VARDIR/tmp con*;
+exec $MYSQL test < $MYSQLTEST_VARDIR/tmp/con@@@.sql;
+exec $MYSQL test < $MYSQLTEST_VARDIR/tmp/con@002fbar.sql;
+exec $MYSQL test < $MYSQLTEST_VARDIR/tmp/con_sch@1ine_gr@1o@1je.sql;
+exec $MYSQL test < $MYSQLTEST_VARDIR/tmp/con@home.sql;
+show tables;
+exec $MYSQL_IMPORT test $MYSQLTEST_VARDIR/tmp/con@@@.txt;
+exec $MYSQL_IMPORT test $MYSQLTEST_VARDIR/tmp/con@002fbar.txt;
+if (`select @@version like '10.6.%'`) {
+# utf8 console output on Windows is fixed in MDEV-26713, until then
+--disable_result_log
+}
+exec $MYSQL_IMPORT test $MYSQLTEST_VARDIR/tmp/con_sch@1ine_gr@1o@1je.txt;
+--enable_result_log
+exec $MYSQL_IMPORT test $MYSQLTEST_VARDIR/tmp/con@home.txt;
+select * from `con_schöne_grüße`;
+select * from `con`;
+select * from `con/bar`;
+select * from `con@home`;
+drop table `con_schöne_grüße`;
+drop table `con`;
+drop table `con/bar`;
+drop table `con@home`;
+
+--echo # End of 10.6 tests
--- a/mysql-test/main/mysqldump.result
+++ b/mysql-test/main/mysqldump.result
@@ -6624,3 +6624,49 @@
 drop view `v'1"2`;
 drop table t1;
 # End of 10.5 tests
+#
+# MDEV-37483 mariadb-dump -T doesn't convert table names
+#
+set names latin1;
+create database foo;
+use foo;
+create table `con_schöne_grüße` (a int) select 1 as a;
+create table `con` (b int) select 2 as b;
+create table `con/bar` (c int) select 3 as c;
+create table `con@home` (d int) select 4 as d;
+drop database foo;
+use test;
+con@002fbar.sql
+con@002fbar.txt
+con@@@.sql
+con@@@.txt
+con@home.sql
+con@home.txt
+con_sch@1ine_gr@1o@1je.sql
+con_sch@1ine_gr@1o@1je.txt
+show tables;
+Tables_in_test
+con
+con/bar
+con@home
+con_schöne_grüße
+test.con: Records: 1  Deleted: 0  Skipped: 0  Warnings: 0
+test.con/bar: Records: 1  Deleted: 0  Skipped: 0  Warnings: 0
+test.con@home: Records: 1  Deleted: 0  Skipped: 0  Warnings: 0
+select * from `con_schöne_grüße`;
+a
+1
+select * from `con`;
+b
+2
+select * from `con/bar`;
+c
+3
+select * from `con@home`;
+d
+4
+drop table `con_schöne_grüße`;
+drop table `con`;
+drop table `con/bar`;
+drop table `con@home`;
+# End of 10.6 tests
